core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -91,7 +90,12 @@ private[spark] class DiskBlockManager(blockManager: BlockManager, conf: SparkCon
   /** List all the files currently stored on disk by the disk manager. */
   def getAllFiles(): Seq[File] = {
     // Get all the files inside the array of array of directories
-    subDirs.flatten.filter(_ != null).flatMap { dir =>
+    subDirs.flatMap { dir =>
```

Comment:
> cloning the array doesn't change any guarantees about whether a call to this get method sees a particular update above or not.

`dir.clone()` is inside the `dir.synchronized` block. It will guarantee that each file in the array either is null, or has finished the construction. If simply reading the array, we may see some files that has not yet finished the construction.
An assumption here is that the item in the `dir` will be set only once.